student/src/main/java/tds/student/web/backing/DialogFrameBacking.java

```diff
@@ -67,18 +74,29 @@ public void setOnLoad ()
   private String getContent (long bankKey, long itemKey) {
     try {
       HttpServletRequest request = HttpContext.getCurrentContext ().getRequest ();
-      StringBuilder urlSB = new StringBuilder ();
-      urlSB.append (request.getRequestURL ().toString ().replace ("DialogFrame.aspx", "API/DialogFrame.axd/getContent"));
-      urlSB.append ("?language=").append (StudentContext.getLanguage ());
-      urlSB.append ("&bankKey=").append (bankKey);
-      urlSB.append ("&itemKey=").append (itemKey);
+      final UriComponentsBuilder uriBuilder = UriComponentsBuilder
+          .fromHttpUrl(request.getRequestURL ().toString ().replace ("DialogFrame.aspx", "API/DialogFrame.axd/getContent"))
+          .queryParam("language", StudentContext.getLanguage())
+          .queryParam("bankKey", bankKey)
+          .queryParam("itemKey", itemKey);
+
+      if (isNotBlank(request.getHeader(X_FORWARDED_HOST))) {
+          uriBuilder.host(request.getHeader(X_FORWARDED_HOST));
+      }
+      if (isNotBlank(request.getHeader(X_FORWARDED_PORT))) {
+          uriBuilder.port(Integer.valueOf(request.getHeader(X_FORWARDED_PORT), 10));
+      }
+      if (isNotBlank(request.getHeader(X_FORWARDED_PROTOCOL))) {
+          uriBuilder.scheme(request.getHeader(X_FORWARDED_PROTOCOL));
+      }
```

Comment:
I worry that there are other places in the legacy code that are also creating new HTTP requests to themselves, but I didn't find any other usages of the GenericRestAPIClient class.